### PR TITLE
Ability to register in system (as user) (#6)

### PR DESCRIPTION
**Pull request by IvanitskyiVolodymyr**

### CODE REVIEW

File: backend/Recodo.API/Controllers/RegisterController.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Recodo.API.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class RegisterController : ControllerBase
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/RegisterController.cs]: Let's use feature controllers, not action (e.g. AuthController)

File: backend/Recodo.API/Controllers/RegisterController.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        private readonly UserService _userService;
+
+        public RegisterController(AuthService authService, UserService userService)
+        {
+            this._authService = authService;
+            this._userService = userService;
+        }
+
+        [HttpPost]
+        public async Task<ActionResult<AuthUserDTO>> Post([FromBody] UserRegisterDTO userDTO)
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/RegisterController.cs]: Name it as function - Register

File: backend/Recodo.API/Recodo.API.csproj [hunk 3]
```diff
@@ -16,12 +17,13 @@
...
     <Folder Include="Controllers\" />
-    <Folder Include="Extensions\" />
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\Recodo.BLL\Recodo.BLL.csproj" />
     <ProjectReference Include="..\Recodo.Common\Recodo.Common.csproj" />
   </ItemGroup>
 
+  <ProjectExtensions><VisualStudio><UserProperties appsettings_1json__JsonSchema="" /></VisualStudio></ProjectExtensions>
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Recodo.API.csproj]: Do you need this?

File: backend/Recodo.API/Extensions/ServiceExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                ValidateIssuer = true,
+                ValidIssuer = jwtAppSettingOptions[nameof(JwtIssuerOptions.Issuer)],
+
+                ValidateAudience = true,
+                ValidAudience = jwtAppSettingOptions[nameof(JwtIssuerOptions.Audience)],
+
+                ValidateIssuerSigningKey = true,
+                IssuerSigningKey = signingKey,
+
+                RequireExpirationTime = false,
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Extensions/ServiceExtensions.cs]: Set it to true and get expiration from appsettings

File: backend/Recodo.BLL/JWT/JwtFactory.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            _jwtSecurityTokenHandler = new JwtSecurityTokenHandler();
+        }
+
+        public async Task<AccessToken> GenerateAccessToken(int id, string userName, string email)
+        {
+            var identity = GenerateClaimsIdentity(id, userName);
+
+            var claims = new[]
+            {
+                 new Claim(JwtRegisteredClaimNames.Sub, userName),
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/JWT/JwtFactory.cs]: Better add id here, we won't use userName

File: backend/Recodo.BLL/MappingProfiles/UserProfile.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using AutoMapper;
+using Recodo.Common.Dtos.User;
+using Recodo.DAL.Entities;
+
+namespace Recodo.BLL.MappingProfiles
+{
+    public class UserProfile:Profile
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/MappingProfiles/UserProfile.cs]: Spaces ' : '

File: backend/Recodo.BLL/Services/AuthService.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            {
+                Token = refreshToken,
+                UserId = id
+            });
+
+            await _context.SaveChangesAsync();
+
+            AccessToken accessToken = await _jwtFactory.GenerateAccessToken(id, userName, userEmail);
+
+            return new AccessTokenDTO(accessToken,refreshToken);
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/AuthService.cs]: If you created refreshToken, then name it TokenDTO

File: backend/Recodo.Common/Auth/AccessToken.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class AccessToken
+    {
+        public string Token { get; }
+        public int ExpiresIn { get; }
+
+        public AccessToken(string token, int expiresIn)
+        {
+            Token = token;
+            ExpiresIn = expiresIn;
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.Common/Auth/AccessToken.cs]: This info is in JWT token itself, you don't need it

File: backend/Recodo.Common/Auth/JwtIssuerOptions.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
+﻿using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Threading.Tasks;
+
+namespace Recodo.Common.Auth
+{
+    public class JwtIssuerOptions
+    {
+        private const int expireTokenTimeInMin = 1440;
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.Common/Auth/JwtIssuerOptions.cs]: Move to appsettings

File: backend/Recodo.API/Controllers/RegisterController.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    [Route("api/[controller]")]
+    [ApiController]
+    public class RegisterController : ControllerBase
+    {
+        private readonly AuthService _authService;
+        private readonly UserService _userService;
+
+        public RegisterController(AuthService authService, UserService userService)
+        {
+            this._authService = authService;
```
Comment (Jyuart) [inline backend/Recodo.API/Controllers/RegisterController.cs]: Please, don't use `this`  😁

File: backend/Recodo.API/Startup.cs [hunk 11]
```diff
@@ -21,25 +22,31 @@ public Startup(IConfiguration configuration)
...
-                .AddFluentValidation(s =>
+            services.AddControllers().AddFluentValidation(s =>
                 {
                     s.RegisterValidatorsFromAssemblyContaining<ModelValidator>();
                 });
+
+            services.AddDbContext<RecodoDbContext>(options =>
+                options.UseNpgsql(Configuration.GetConnectionString("DefaultConnection")));
+
+            //Register your mappers in Recodo.API.Extensions.ServiceExtensions.cs
```
Comment (Jyuart) [inline backend/Recodo.API/Startup.cs]: I think you can once register them from assembly and forget about this registering
https://docs.automapper.org/en/stable/Configuration.html#assembly-scanning-for-auto-configuration

File: backend/Recodo.Common/Dtos/User/UserRegisterDTO.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recodo.Common.Dtos.User
+{
+    public class UserRegisterDTO:UserDTO
```
Comment (Jyuart) [inline backend/Recodo.Common/Dtos/User/UserRegisterDTO.cs]:
```suggestion
    public class NewUserDto : UserDTO
```

Maybe something like this?

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by Jyuart — APPROVED** [review]

**Pull request merged**